### ISSUE TITLE
hyperparam_search.py

### ISSUE THREAD

**Issue by karthik78437** [karthik78437/karthik#7]
"""
hyperparam_search.py

Example hyperparameter search using Keras Tuner (RandomSearch) if available.
If Keras Tuner is not installed, fallback to a grid/loop search.

This script demonstrates searching over:
 - lstm_units
 - dropout
 - learning_rate
 - batch_size
 - bidirectional
"""

import os
import numpy as np
from data_generation import generate_complex_timeseries, create_sequences
from utils import train_test_split_sequences, scale_sequences
from models import build_attention_lstm, build_baseline_lstm

# Optional import for keras-tuner:
try:
    import kerastuner as kt
    KERASTUNER_AVAILABLE = True
except Exception:
    KERASTUNER_AVAILABLE = False

import tensorflow as tf

def simple_grid_search(X_tr, y_tr, X_val, y_val, input_shape, model_type='attention', grid=None, results_dir='hp_results'):
    """
    Simple exhaustive grid search (use for small grid only).
    """
    os.makedirs(results_dir, exist_ok=True)
    if grid is None:
        grid = {
            'lstm_units': [32, 64],
            'dropout': [0.1, 0.3],
            'batch_size': [64, 128]
        }

    import itertools
    best_cfg = None
    best_val = float('inf')

    for lstm_units, dropout, batch_size in itertools.product(grid['lstm_units'], grid['dropout'], grid['batch_size']):
        print("Testing:", lstm_units, dropout, batch_size)
        if model_type == 'baseline':
            model = build_baseline_lstm(input_shape, lstm_units=lstm_units, dropout=dropout)
        else:
            model = build_attention_lstm(input_shape, lstm_units=lstm_units, dropout=dropout)

        # compile optimizer with small LR
        model.compile(optimizer=tf.keras.optimizers.Adam(learning_rate=1e-3), loss='mse', metrics=['mae'])
        history = model.fit(X_tr, y_tr, validation_data=(X_val, y_val), epochs=20, batch_size=batch_size, verbose=0)
        val_loss = min(history.history['val_loss'])
        print("Val loss:", val_loss)
        # Save if improved
        if val_loss < best_val:
            best_val = val_loss
            best_cfg = {'lstm_units': lstm_units, 'dropout': dropout, 'batch_size': batch_size}
    print("Best:", best_cfg, "val_loss:", best_val)
    return best_cfg, best_val

if __name__ == '__main__':
    # generate and split just for search
    df = generate_complex_timeseries(n_steps=1200, n_exog=3)
    X, y = create_sequences(df, seq_len=48)
    X_tr_full, X_test, y_tr_full, y_test = train_test_split_sequences(X, y, test_size=0.2)
    # split train-> train/val
    split = int(0.8 * len(X_tr_full))
    X_tr = X_tr_full[:split]; y_tr = y_tr_full[:split]
    X_val = X_tr_full[split:]; y_val = y_tr_full[split:]
    X_tr_s, X_val_s, _ = scale_sequences(X_tr, X_val)
    input_shape = (X_tr_s.shape[1], X_tr_s.shape[2])

    best_cfg, best_val = simple_grid_search(X_tr_s, y_tr, X_val_s, y_val, input_shape, model_type='attention')
    print("Done.")
